File: ios/AppleAdsAttribution.m
#import "AppleAdsAttribution.h"
#import <React/RCTLog.h>
#import <AdServices/AdServices.h>

@implementation AppleAdsAttribution
static NSString *const RNAAAErrorDomain = @"RNAAAErrorDomain";
static int NUM_RETRIES = 3;

RCT_EXPORT_MODULE();

- (dispatch_queue_t)methodQueue
{
    return dispatch_get_main_queue();
}

+ (void)rejectPromiseWithNSError:(RCTPromiseRejectBlock)reject error:(NSError * _Nullable)error {
    
    if (error == NULL) {
        reject(@"unknown", @"Failed with unknown error", nil);
    } else {
        NSMutableDictionary *userInfo = [NSMutableDictionary dictionary];
        [userInfo setValue:error forKey:@"message"];
        [userInfo setValue:@(error.code) forKey:@"nativeErrorCode"];
        NSError *newErrorWithUserInfo = [NSError errorWithDomain:RNAAAErrorDomain
                                                            code:100
                                                        userInfo:userInfo];
        reject(@"unknown", error.localizedDescription, newErrorWithUserInfo);
    }
}

+ (void)rejectPromiseWithUserInfo:(RCTPromiseRejectBlock)reject userInfo:(NSMutableDictionary *)userInfo {
    
    NSError *error = [NSError errorWithDomain:RNAAAErrorDomain code:100 userInfo:userInfo];
    reject(userInfo[@"code"], userInfo[@"message"], error);
}

+ (BOOL)isSimulator {
    #if (TARGET_OS_SIMULATOR)
    return YES;
    #else
    return NO;
    #endif
}

/**
 * Uses the provided token to request attribution data from apples AdServices API.
 */
+ (void) requestAdServicesAttributionDataUsingToken:(NSString *) token
                                        retriesLeft:(int) retriesLeft
                                  completionHandler:(void (^)(NSDictionary * _Nullable data, NSError * _Nullable error))completionHandler
API_AVAILABLE(ios(14.3)) {
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
    [request setHTTPMethod:@"POST"];
    [request setValue:@"text/plain" forHTTPHeaderField:@"Content-Type"];
    [request setURL:[NSURL URLWithString:@"https://api-adservices.apple.com/api/v1/"]];
    [request setHTTPBody:[token dataUsingEncoding:NSUTF8StringEncoding]];
    
    [[[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable reqError) {
        
        // Status codes like 404 doesn't generate an error, so check that request was successful by making sure it's a 200 code
        if ([response isKindOfClass:[NSHTTPURLResponse class]]) {
            NSInteger statusCode = [(NSHTTPURLResponse *)response statusCode];
            if (statusCode != 200) {
                if ((statusCode == 404 || statusCode == 500) && retriesLeft > 0) {
                    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                        [AppleAdsAttribution requestAdServicesAttributionDataUsingToken:token retriesLeft:retriesLeft-1 completionHandler:completionHandler];
                    });
                } else {
                    NSMutableDictionary* details = [NSMutableDictionary dictionary];
                    [details setValue:[NSString stringWithFormat:@"Request to get data from Adservices API failed with status code %ld. Re-tried %i times", (long)statusCode, NUM_RETRIES - retriesLeft] forKey:NSLocalizedDescriptionKey];
                    NSError* error = [NSError errorWithDomain:RNAAAErrorDomain code:100 userInfo:details];
                    completionHandler(nil, error);
                }
                return;
            }
        }
        
        if (reqError != nil) {
            completionHandler(nil, reqError);
        } else if (data) {
            NSError* serializationError = nil;
            NSDictionary* attributionDataDictionary = [NSJSONSerialization JSONObjectWithData:data options:0 error:&serializationError];
            if (!serializationError && attributionDataDictionary) {
                completionHandler(attributionDataDictionary, nil);
            } else {
                completionHandler(nil, serializationError);
            }
        } else {
            // No error and no data, not sure if it can happen..
            NSMutableDictionary* details = [NSMutableDictionary dictionary];
            [details setValue:@"Request to Adservices API failed with unknown error" forKey:NSLocalizedDescriptionKey];
            NSError* error = [NSError errorWithDomain:RNAAAErrorDomain code:100 userInfo:details];
            completionHandler(nil, error);
            
        }
    }] resume];
}

/**
 * Tries to generate an attribution token that then can be used for calls to apples AdServices API.
 * Returns nil if token couldn't be generated.
 */
+ (NSString *) getAdServicesAttributionToken:(NSError * _Nullable *)error {
    if([AppleAdsAttribution isSimulator]) {
        if (error != NULL) {
            NSMutableDictionary* details = [NSMutableDictionary dictionary];
            [details setValue:@"Error getting token, not available in Simulator" forKey:NSLocalizedDescriptionKey];
            *error = [NSError errorWithDomain:RNAAAErrorDomain code:100 userInfo:details];
        }
        return nil;
    }

    if (@available(iOS 14.3, *))
    {
        Class AAAttributionClass = NSClassFromString(@"AAAttribution");
        if (AAAttributionClass) {
            NSString *attributionToken = [AAAttributionClass attributionTokenWithError:error];
            if (*error == nil && attributionToken) {
                return attributionToken;
            }
        } else {
            NSMutableDictionary* details = [NSMutableDictionary dictionary];
            [details setValue:@"Error getting token, AAAttributionClass not found" forKey:NSLocalizedDescriptionKey];
            *error = [NSError errorWithDomain:RNAAAErrorDomain code:100 userInfo:details];
        }
    } else if (error != NULL) {
        NSMutableDictionary* details = [NSMutableDictionary dictionary];
        [details setValue:@"Error getting token, AdServices not available pre iOS 14.3" forKey:NSLocalizedDescriptionKey];
        *error = [NSError errorWithDomain:RNAAAErrorDomain code:100 userInfo:details];
    }
    return nil;
}

/**
 * Generates an attributionToken that it then uses to request attribution data from apples AdServices API.
 * Returns and error if attribution data couldn't be fetched
 */
+ (void) getAdServicesAttributionDataWithCompletionHandler: (void (^)(NSDictionary * _Nullable data, NSError * _Nullable error))completionHandler {
    
    if (@available(iOS 14.3, *))
    {
        NSError *tokenError = nil;
        NSString* attributionToken = [AppleAdsAttribution getAdServicesAttributionToken:&tokenError];
        
        if (attributionToken) {
            [AppleAdsAttribution requestAdServicesAttributionDataUsingToken:attributionToken retriesLeft:NUM_RETRIES completionHandler:completionHandler];
        } else {
            // No token
            completionHandler(nil, tokenError);
        }
    } else {
        // Not supported on this device
        NSMutableDictionary* details = [NSMutableDictionary dictionary];
        [details setValue:@"AdServices not available pre iOS 14.3" forKey:NSLocalizedDescriptionKey];
        NSError* error = [NSError errorWithDomain:RNAAAErrorDomain code:100 userInfo:details];
        completionHandler(nil, error);
    }
}

/**
 * Tries to get attribution data first using the AdServices API.
 * Rejected with error if both fails
 */
RCT_EXPORT_METHOD(getAttributionData:
                  (RCTPromiseResolveBlock) resolve
                  rejecter:
                  (RCTPromiseRejectBlock) reject) {
    
    [AppleAdsAttribution getAdServicesAttributionDataWithCompletionHandler:^(NSDictionary * _Nullable attributionData, NSError * _Nullable adServicesError) {
        if (attributionData != nil) {
            resolve(attributionData);
        } else {
            // Reject with both error messages
            NSString *combinedErrorMessage = [NSString stringWithFormat:@"Ad services error: %@.", adServicesError != NULL ? adServicesError.localizedDescription : @"no error message"];
            
            [AppleAdsAttribution rejectPromiseWithUserInfo:reject
                                                  userInfo:[@{
                                                    @"code" : @"unknown",
                                                    @"message" : combinedErrorMessage
                                                  } mutableCopy]];
        }
    }];
}

/**
 * Tries to generate an attribution token that then can be used for calls to Apples AdServices API.
 * Rejected with error if token couldn't be generated.
 */
RCT_EXPORT_METHOD(getAdServicesAttributionToken: (RCTPromiseResolveBlock) resolve rejecter: (RCTPromiseRejectBlock) reject) {
    NSError *error = nil;
    NSString* attributionToken = [AppleAdsAttribution getAdServicesAttributionToken:&error];
    
    if (attributionToken != nil) {
        resolve(attributionToken);
    } else {
        [AppleAdsAttribution rejectPromiseWithNSError:reject error:error];
    }
}

/**
 * Tries to get attribution data from apples AdServices API.
 * Rejected with error if data couldn't be fetched.
 */
RCT_EXPORT_METHOD(getAdServicesAttributionData: (RCTPromiseResolveBlock) resolve rejecter: (RCTPromiseRejectBlock) reject) {
    [AppleAdsAttribution getAdServicesAttributionDataWithCompletionHandler:^(NSDictionary * _Nullable attributionData, NSError * _Nullable error) {
        if (attributionData != nil) {
            resolve(attributionData);
        } else {
            [AppleAdsAttribution rejectPromiseWithNSError:reject error:error];
        }
    }];
}

@end

File: ios/AppleAdsAttribution.swift
//
//  AppleAdsAttribution.swift
//  AppleAdsAttribution
//
//  Created by Garry on 17/09/24.
//  Copyright © 2024 Facebook. All rights reserved.
//

import Foundation
import AdServices
import React

// Error domain used for rejection errors
let RNAAAErrorDomain = "RNAAAErrorDomain"
let NUM_RETRIES = 3

@objc(AppleAdsAttribution)
class AppleAdsAttribution: NSObject, RCTBridgeModule {

  // MARK: - Queue

  static var methodQueue: DispatchQueue {
    return DispatchQueue.main
  }

  // MARK: - Error Handling

  static func rejectPromise(with error: Error?, using reject: @escaping RCTPromiseRejectBlock) {
    if let error = error {
      reject("unknown", error.localizedDescription, error)
    } else {
      reject("unknown", "Failed with unknown error", nil)
    }
  }

  static func rejectPromise(with userInfo: [String: Any], using reject: @escaping RCTPromiseRejectBlock) {
    let error = NSError(domain: RNAAAErrorDomain, code: 100, userInfo: userInfo)
    reject(userInfo["code"] as? String ?? "", userInfo["message"] as? String ?? "", error)
  }

  // MARK: - Simulator Check

  static func isSimulator() -> Bool {
    #if targetEnvironment(simulator)
      return true
    #else
      return false
    #endif
  }

  // MARK: - AdServices API Calls

  @available(iOS 14.3, *)
  static func requestAdServicesAttributionData(using token: String, retriesLeft: Int, completion: @escaping (NSDictionary?, Error?) -> Void) {
    var urlRequest = URLRequest(url: URL(string: "https://api-adservices.apple.com/api/v1/")!)
    urlRequest.httpMethod = "POST"
    urlRequest.setValue("text/plain", forHTTPHeaderField: "Content-Type")
    urlRequest.httpBody = token.data(using: .utf8)

    let dataTask = URLSession.shared.dataTask(with: urlRequest) { data, response, error in
      guard let response = response as? HTTPURLResponse else { return }

      // Check for successful status code (200)
      if response.statusCode != 200 {
        if (response.statusCode == 404 || response.statusCode == 500) && retriesLeft > 0 {
          DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(5)) {
            requestAdServicesAttributionData(using: token, retriesLeft: retriesLeft - 1, completion: completion)
          }
        } else {
          let details = ["NSLocalizedDescriptionKey": String(format: "Request to get data from Adservices API failed with status code %ld. Re-tried %i times", response.statusCode, NUM_RETRIES - retriesLeft)]
          let error = NSError(domain: RNAAAErrorDomain, code: 100, userInfo: details)
          completion(nil, error)
        }
        return
      }

      if let error = error {
        completion(nil, error)
        return
      }

      guard let data = data else {
        let details = ["NSLocalizedDescriptionKey": "Request to Adservices API failed with unknown error"]
        let error = NSError(domain: RNAAAErrorDomain, code: 100, userInfo: details)
        completion(nil, error)
        return
      }

      do {
        let attributionData = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
        completion(attributionData, nil)
      } catch {
        completion(nil, error)
      }
    }
    dataTask.resume()
  }

  // MARK: - AdServices Token Generation

  static func getAdServicesAttributionToken() -> String? {
    if isSimulator() {
      let details = ["NSLocalizedDescriptionKey": "Error getting token, not available in Simulator"]
      let error = NSError(domain: RNAAAErrorDomain, code: 100, userInfo: details)
      return nil
    }

    if #available(iOS 14.3, *) {
      let AAAttributionClass = NSClassFromString("AAAttribution")
      if let AAAttributionClass = AAAttributionClass {
        do {
          let attributionToken = try AAAttributionClass.attributionToken()
          return attributionToken
        } catch {
          let details = ["NSLocalizedDescriptionKey": "Error getting token, AAAttributionClass not found"]
          let error = NSError(domain: RNAAAErrorDomain, code: 100, userInfo: details)
          return nil
        }
      } else {
        let details = ["NSLocalizedDescriptionKey": "Error getting token, AAAttributionClass not found"]
        let error = NSError(domain: RNAAAErrorDomain, code: 100, userInfo: details)
        return nil
      }
    } else {
      let details = ["NSLocalizedDescriptionKey": "Error getting token, AdServices not available pre iOS 14.3"]
      let error = NSError(domain: RNAAAErrorDomain, code: 100, userInfo: details)
      return nil
    }
  }

  // MARK: - React Native Module Methods

/**
 * Tries to get attribution data first using the AdServices API. If it fails it fallbacks to the old iAd API.
 * Rejected with error if both fails
 */
  @objc(getAttributionData)
  func getAttributionData(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    if #available(iOS 14.3, *) {
      getAdServicesAttributionDataWithCompletionHandler { attributionData, adServicesError in
        if let attributionData = attributionData {
          resolve(attributionData)
        } else {
          getiAdAttributionDataWithCompletionHandler { data, iAdError in
            if let data = data {
              resolve(data)
            } else {
              let combinedErrorMessage = "Ad services error: \(adServicesError?.localizedDescription ?? "no error message"). \niAD error: \(iAdError?.localizedDescription ?? "no error message")"
              rejectPromise(with: ["code": "unknown", "message": combinedErrorMessage], using: reject)
            }
          }
        }
      }
    } else {
      getiAdAttributionDataWithCompletionHandler { data, error in
        if let data = data {
          resolve(data)
        } else {
          rejectPromise(with: error, using: reject)
        }
      }
    }
  }

/**
 * Tries to get attribution data using the old iAd API.
 * Rejected with error if it failed to get data
 *  */
  @objc(getiAdAttributionData)
  func getiAdAttributionData(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    getiAdAttributionDataWithCompletionHandler { data, error in
      if let data = data {
        resolve(data)
      } else {
        rejectPromise(with: error, using: reject)
      }
    }
  }

/**
 * Tries to generate an attribution token that then can be used for calls to Apples AdServices API.
 * Rejected with error if token couldn't be generated.
 */
  @objc(getAdServicesAttributionToken)
  func getAdServicesAttributionToken(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    if let attributionToken = getAdServicesAttributionToken() {
      resolve(attributionToken)
    } else {
      rejectPromise(with: error, using: reject)
    }
  }

/**
 * Tries to get attribution data from apples AdServices API.
 * Rejected with error if data couldn't be fetched.
 */
  @objc(getAdServicesAttributionData)
  func getAdServicesAttributionData(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    getAdServicesAttributionDataWithCompletionHandler { attributionData, error in
      if let attributionData = attributionData {
        resolve(attributionData)
      } else {
        rejectPromise(with: error, using: reject)
      }
    }
  }

  // MARK: - iAd API Calls

/**
 * Gets attribution data from the old iAd API.
 * completionHandler will return nil with an error if attribution data couldn't be retrieved. Reasons for failing may be that the user disabled tracking or that the iOS version is < 10.
 */
  static func getiAdAttributionDataWithCompletionHandler(completion: @escaping (NSDictionary?, Error?) -> Void) {
    if #available(iOS 10, *) {
      ADClient.shared.requestAttributionDetails { attributionDetails, error in
        if let error = error {
          completion(nil, error)
        } else {
          completion(attributionDetails, nil)
        }
      }
    } else {
      let details = ["NSLocalizedDescriptionKey": "iAd ADClient not available"]
      let error = NSError(domain: RNAAAErrorDomain, code: 100, userInfo: details)
      completion(nil, error)
    }
  }
}
